.travis.yml

```diff
@@ -80,86 +65,87 @@ install:
-    cargo-deadlinks -V | grep $DEADLINKS_VERS || cargo install cargo-deadlinks --vers $DEADLINKS_VERS --force
-    fi
-- |
-  if [[ "$FEATURE" == "fmt" ]]; then
-    cargo-audit -V || cargo install cargo-audit --force
-    rustfmt -V | grep $RUSTFMT_VERS || cargo install rustfmt --vers $RUSTFMT_VERS --force
-    nvm install 8
-    nvm use 8
-    npm install -g cspell
-    cspell --version
-    npm install -g markdownlint-cli
-    markdownlint --version
-    fi
-- |
-  case "$FEATURE" in
-  clippy* )
-    cargo clippy --version | grep $CLIPPY_VERS || cargo install clippy --force --vers $CLIPPY_VERS
-  ;;
-  esac
-- cargo install --list
 
-script: |
-  case "$FEATURE" in
+jobs:
+  include:
+  # Formatting & other lints that do not require compilation
+  - env: FEATURE=fmt
+    install:
+    - cargo-audit -V || cargo install cargo-audit --force
+    - rustfmt -V | grep $RUSTFMT_VERS || cargo install rustfmt --vers $RUSTFMT_VERS --force
+    - nvm install 8 && nvm use 8
+    - npm install cspell
+    - ./node_modules/.bin/cspell --version
+    - npm install markdownlint-cli
+    - ./node_modules/.bin/markdownlint --version
+    script:
+    - cd exonum && cargo fmt -- --write-mode=diff && cd .. &&
+    - cd sandbox && cargo fmt -- --write-mode=diff && cd .. &&
+    - cd testkit && cargo fmt -- --write-mode=diff && cd .. &&
+    - cd services/time && cargo fmt -- --write-mode=diff && cd ../.. &&
+    - ./node_modules/.bin/cspell sandbox/{src,examples,tests}/**/*.rs &&
+    - ./node_modules/.bin/cspell exonum/{src,benches,tests}/**/*.rs &&
+    - ./node_modules/.bin/cspell testkit/{src,examples,tests}/**/*.rs &&
+    - ./node_modules/.bin/cspell services/time/{src,examples,tests}/**/*.rs &&
+    - find . -not -path "./3rdparty/*" -name "*.md" | xargs markdownlint --config .markdownlintrc
+
+  # Clippy linting
+  - &clippy-core
+    env: FEATURE=clippy-core
+    rust: nightly-2018-01-22
+    install:
+    - cargo clippy --version | grep $CLIPPY_VERS || cargo install clippy --force --vers $CLIPPY_VERS
+    script:
+    - cd exonum && cargo clippy --verbose --features long_benchmarks -- -D warnings
+  - <<: *clippy-core
+    env: FEATURE=clippy-sandbox
+    script:
+    - cd sandbox && cargo clippy --verbose -- -D warnings
+  - <<: *clippy-core
+    env: FEATURE=clippy-testkit
+    script:
+    - cd testkit && cargo clippy --verbose -- -D warnings
+  - <<: *clippy-core
+    env: FEATURE=clippy-time
+    script:
+    - cd services/time && cargo clippy --verbose -- -D warnings
+
+  # Tests
+  - env: FEATURE=test-core
+    script:
+    - cargo test -p exonum --verbose --tests --lib --features "float_serialize"
+  - env: FEATURE=test-sandbox
+    script:
+    - cargo test -p sandbox --verbose
+  - env: FEATURE=test-testkit
+    script:
+    - cargo test -p exonum-testkit --verbose
+    - cargo run -p exonum-testkit --example timestamping
+    - cargo run -p exonum-testkit --example configuration_change
+  - env: FEATURE=test-time
+    script:
```

Comment:
Are we sure that we need separate `FEATURE=test-core`, `FEATURE=test-testkit`, etc.?

With multiple features, we'll be building exonum itself several times for each job. With `cargo test --all`, the total amount of work will be smaller.

Reply:
👍 . @SerhiiOryshych Could you consolidate the present jobs a single test job and a single clippy job to measure the impact on the build time? You can use `test-core` / `clippy-core` for `FEATURE` values in order to reuse existing caches.